### ISSUE TITLE
Process multiline logs from multiple docker images

### ISSUE THREAD

**Issue by kafkapre** [elastic/logstash#4308]
I run docker container with gelf driver and would like to collapse multiline logs in Logstash. My Logstash conf. 

```
input {
    gelf {} 
}
filter{
    multiline {
      pattern => "^%{TIMESTAMP_ISO8601}"
      negate => true
      what => "previous"
      source => "short_message"
      }
}
output {
    stdout { codec => rubydebug }
}
```

It works perfectly when I process logs from one docker container, but for two or more it does not work, because it collapse messages of both (or more) logs streams. 

I would expect, that setting up multilining in input would solve the problem.

```
input {
    gelf {      
         codec => multiline {
            pattern => "^%{TIMESTAMP_ISO8601}"
            negate => true
            what => "previous"         
     }
}
```

but multilining does not work correctly with this set up (seems because of [bug](https://github.com/logstash-plugins/logstash-input-gelf/pull/18)). Any suggestions? Thanks.

I am using: Docker 1.9.1, Logstash 2.1

**Comment by purbon** [comment]
multiline filter is not thread safe, so is not good to collapse logs as you described. one improvement to this would be to use the multiline codec in your input section, see https://www.elastic.co/guide/en/logstash/current/plugins-codecs-multiline.html for details.

Closing this as is a known problem of the multiline filter, feel free to reopen any other issue you might find. Keep in mind https://discuss.elastic.co/, a great source of community support.

**Comment by kafkapre** [comment]
Exactly, I would prefer to put multiline into input section. But, as I described above, it does not work correctly.

**Comment by purbon** [comment]
Use the codec syntax, your example is wrong at my understanding.

/purbon

**Comment by kafkapre** [comment]
Oh, sorry I wrote i incorrectly.  This is correct (I have also corrected it above).
```
input {
    gelf {      
         codec => multiline {
            pattern => "^%{TIMESTAMP_ISO8601}"
            negate => true
            what => "previous"         
     }
}
```

And for this case, multilining does not work.

**Comment by jordansissel** [comment]
Hmm, I think this is a problem that the gelf input needs to get together with the IdentityMapCodec. Thoughts, @guyboertje?

**Comment by guyboertje** [comment]
I will look into it

**Comment by purbon** [comment]
At my understanding, after looking at the code base https://github.com/logstash-plugins/logstash-input-gelf/blob/master/lib/logstash/inputs/gelf.rb, the gelf codec have the codec directive, but the codec is never used within. So this is the bug at my understanding.

@kafkapre I agree there is a bug, no codec is actually working there, can you please reopen this issue in the https://github.com/logstash-plugins/logstash-input-gelf repo? thanks a lot for your finding.

**Comment by uschtwill** [comment]
For others that might be searching for a solution to this. This worked beautifully for me: https://stackoverflow.com/questions/34075538/elk-process-multiline-logs-from-multiple-docker-images

**Comment by guyboertje** [comment]
This plugin does not use a codec because it receives a JSON string that represents an event. Codecs are meant to take a line and create an Event. Filters are meant to operate on Events. The multiline codec buffers lines until some condition is seen - it would have to be changed to buffer events and match on a field, but wait - that is what the Mulitline Filter does. So we would need a way to run the multiline filter after an input but before the queue. We can't do that at the moment.

**Comment by jmreicha** [comment]
@kafkapre did you ever find a workaround?

**Comment by guyboertje** [comment]
@jmreicha and other future readers:

The solution offered in the Stackoverflow link from @uschtwill **IS** the solution for this issue. There are no workarounds. There may well be a performance impact when using the multiline filter because, by design, it needs to receive every event. This is so because we can't ensure that events with the same identity will always travel down the same path in parallel multithreaded filter stages.

**Comment by jmreicha** [comment]
@guyboertje Thanks for the update.  I was worried about performance but maybe it's okay?  Do you have a ballpark of how many events a single thread can handle?

**Comment by guyboertje** [comment]
@jmreicha - unfortunately, the answer to that question is very difficult to answer because it depends on the version of LS and the complexity of the configuration i.e. whether you have other filters before or after the multiline filter, the rate at which the events are 'pumped' in via the input(s) and the time taken for the output to deliver an event out.

This is a [fantastic presentation](http://www.slideshare.net/avleenvig/elk-mooseively-scaling-your-log-system) from Avleen at Etsy that give v good advice on systematic tuning of the Elastic Stack.

**Comment by jmreicha** [comment]
I'll check it out, thanks.

**Comment by mathiasbn** [comment]
I'm a bit confused that this issue is closed. 
- The proposed solution (the link to stackoverflow) has performance implications.
- The multiline filter is deprecated
- For some reason the multiline filter cannot be found on logstash:5.0.2-1

**Comment by antdavidl** [comment]
I'm in a similar scenario in which I'm being unable to make the _multiline_ codec works properly with the _gelf_ plug-in in ELK 5.x

The solution from @stream_identity is using the _multiline_ filter with the _stream_identity_ parameters, but the _multiline_ filter is deprecated and unavailable now, so I think we are now at the starting point :-(

**Comment by maxblaze** [comment]
Bump. This *really* needs to be reopened.

**Comment by danielmotaleite** [comment]
Logstash is the wrong place to fix multiline events, you need to fix as soon as possible, so inside the app or just after the docker:

https://github.com/elastic/beats/issues/918

I hope someone implements the new docker logger beat plugin with multiline support, as in filebeat

**Comment by maxblaze** [comment]
@danielmotaleite I completely disagree. One of the original killer features of Logstash was the ability to handle multiline events.

**Comment by jordansissel** [comment]
@maxblaze I'll reopen this.

For folks curious about the future, there's some problems to discuss:

1) The multiline filter is gone and probably not coming back. For performance, Logstash will process things out of order (when using multiple pipeline worker threads), which means multiline filtering, even if present, would harm performance. 
2) The aggrregate filter can do what the multiline filter did. It has the same single-threaded, ordered-processing limitations as the multiline filter did.
3) It could be possible to add codec support to the gelf input, but it is unclear if this will solve y'all's problems.

**Comment by danielmotaleite** [comment]
With filebeat being able to automatically process docker logs (via json-file plugin) and it can manage and merge the multiline in the multiple json events, i would say again that trying to do multiline in logstash is a lost battle... either you can only use one thread to get in-order logs so it can work, or you will get sooner or badly merged log lines or lines that are not merged. The more multi-line events you have and the more load you have, the higher is the risk of broken logs.

the only sane way to solve the multi-line is closer to the source as possible. If not possible in the app producing the logs, the next guy (filebeat, fluent, etc) should be the one trying to do

**Comment by Shokodemon** [comment]
jesus christ how is this still open

**Comment by jordansissel** [comment]
@Shokodemon There is no need for abusive language.

filebeat is the recommended solution for what you are doing. https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-input-docker.html

**Comment by Shokodemon** [comment]
@jordansissel fixed. and thank you. luckily, i have a filebeat setup at hand that i've prepared before.

also thank you for the emotional circuit breaker nudge.

**Comment by jordansissel** [comment]
@Shokodemon :+1: I hope filebeat gets you on a path to success! Let us know (probably in elastic/beats or on discuss.elastic.co) if you have issues doing multiline w/ filebeat for docker logs.

**Comment by jhmartin** [comment]
Per https://github.com/moby/moby/issues/17763 , docker considers the json-files as private and won't support changes that make it easier to read them. This makes gelf logging attractive as it doesn't require fighting dockerd or running filebeat as privileged.

**Comment by dino-daniel-yu** [comment]
So if the input type codec is useless for the gelf plugin in logstash. Shouldn't it then be removed from the docs https://www.elastic.co/guide/en/logstash/6.7/plugins-inputs-gelf.html?

Does anyone have experience, recommendations or current tutorials on how to efficiently aggregate these separate Docker log fragments using Filebeat or Logstash?
